File: LDA/SCVB0.py
'''
Created on Mar 4, 2014

@author: vspathak
'''
import numpy
import math


class SCVB0(object):
    
    ''' Implementation of Stochastic CVB0'''
    def initParam(self):
        self.w = 12419
        self.k = 10
        self.d = 10
        self.nPhi = numpy.zeros((self.w, self.k), float)
        self.nTheta = numpy.zeros((self.d+1, self.k), float)
        self.nz = numpy.zeros((self.k), float)
        self.s = 10
        self.tau = 1000
        self.kappa = 0.9
        self.t = 1
        self.rhoPhi = self.s / math.pow((self.tau + self.t), self.kappa)
        self.rhoTheta = self.s / math.pow((self.tau + self.t), self.kappa)
        self.alpha = numpy.empty(self.k)
        self.alpha.fill(0.1)  # 0.1
        self.eta = numpy.empty(self.w)
        self.eta.fill(0.01)  # 0.01
        self.c = 467714
        self.Cj = {}
        self.M = 0
        '''for each mini-batch'''
        self.nPhiHat = numpy.zeros((self.w, self.k), float)
        self.nzHat = numpy.zeros((self.k), float)
        
        self.gamma = numpy.zeros((self.w, self.k), float)
    
    def run(self, docList):
        self.M = 0
        for doc in docList:
            temp = 0
            for term in doc.termDict :
                temp += doc.termDict[term]
            self.Cj[doc.id] = temp
            self.M += temp
        j = 0
        for doc in docList:
            j += 1
            i = -1
            for term in doc.termDict :
                i += 1
                self.updateGamma(i, j, term)
                self.updateNTheta(i, j, term)
            i = -1
            for term in doc.termDict :
                i += 1
                self.updateGamma(i, j, term)
                self.updateNTheta(i, j, term)
                for k in range(0, self.k - 1):
                    self.nPhiHat[term][k] += self.nPhiHat[term][k] + self.c * self.gamma[term][k]
                    self.nzHat[k] += self.nzHat[k] + self.c * self.gamma[term][k]
        self.updateNPhi()
        self.updateNZ()
        print self.nTheta
        print "\n"
        print self.nPhi[4496] ,self.nPhi[4497]  
    
    def updateGamma(self, i, j, Wij):
        # Equation 5
        sigmaEta = 0.0
        for e in self.eta:
            sigmaEta += e
        for k in range(0, self.k-1):
            self.gamma[Wij][k] = ((self.nPhi[Wij][k] + self.eta[Wij]) / (self.nz[k] + sigmaEta)) * (self.nTheta[j][k] + self.alpha[k])
        return

    def updateNTheta(self, i, j, Wij):
        # Equation 9 instead of Equation 6, we are using "clumping"
        for k in range(0, self.k-1):
            self.nTheta[j][k] = ((1 - self.rhoTheta) * self.nTheta[j][k]) + (self.rhoTheta * self.Cj[j] * self.gamma[Wij][k]) 
        return

    def updateNPhi(self):
        # Equation 7
        for k in range(0, self.k-1):
            for w in range(1, self.w):
                self.nPhiHat[w][k] = (self.c * self.gamma[w][k]) / self.M
                self.nPhi[w][k] = ((1 - self.rhoPhi) * self.nPhi[w][k]) + (self.rhoPhi * self.nPhiHat[w][k]) 
        return
    
    def updateNZ(self):
        # Equation 8
        for k in range(0, self.k-1):
            for w in range(1, self.w):
                self.nzHat[k] += self.gamma[w][k]
            self.nzHat[k] = (self.c * self.nzHat[k]) / self.M
            self.nz[k] = ((1 - self.rhoPhi) * self.nz[k]) + (self.rhoPhi * self.nzHat[k])  
        return
            
    def __init__(self):
        self.initParam()
        

File: TopicModeling/LDA/SCVB0.cpp
/*
 * SCVB0.cpp
 *
 *  Created on: Mar 7, 2014
 *      Author: Varad Pathak, Arul Samuel
 */

#include "SCVB0.h"
#include <math.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <algorithm>
#include "MiniBatch.h"

using namespace std;

SCVB0::SCVB0(int iter, int numberOfTopics, int vocabSize, int numOfDocs,
		int corpusSize) {
	iterations = iter;
	K = numberOfTopics;
	W = vocabSize;
	D = numOfDocs;
	C = corpusSize;

	s = 10;
	tau = 1000;
	kappa = 0.9;
	t = 1;

	alpha = 0.1;
	eta = 0.01;

	rhoPhi = s / pow((tau + t), kappa);
	rhoTheta = s / pow((tau + t), kappa);

	nPhi = new double*[W + 1];
	nTheta = new double*[D + 1];
	nz = new double[K];
	memset(nz, 0, sizeof(nz));

	for (int w = 0; w < W + 1; w++) {
		nPhi[w] = new double[K];
		for (int k = 0; k < K; ++k) {
			nPhi[w][k] = ((double) (rand() % (W * K))) / (W * K);
			nz[k] += nPhi[w][k];
		}
	}
	for (int d = 0; d < D + 1; d++) {
		nTheta[d] = new double[K];
		for (int k = 0; k < K; ++k) {
			nTheta[d][k] = ((double) (rand() % (D * K))) / (D * K);
		}
	}
}

SCVB0::~SCVB0() {
	for (int w = 0; w < W + 1; w++) {
		delete[] (nPhi[w]);
	}
	for (int d = 0; d < D + 1; d++) {
		delete[] (nTheta[d]);
	}
	delete[] (nPhi);
	delete[] (nTheta);
	delete[] (nz);
}

void SCVB0::run(MiniBatch miniBatch) {
	vector<Document> docVector = *miniBatch.docVector;
//	cout << "MiniBatchSize: " << miniBatch.M << endl;

	double **nPhiHat = new double*[W + 1];
	double *nzHat = new double[K];
	double **gamma = new double*[W + 1];

	for (int w = 0; w < W + 1; w++) {
		nPhiHat[w] = new double[K];
		gamma[w] = new double[K];
		memset(nPhiHat[w], 0, sizeof(nPhiHat[w]));
		memset(gamma[w], 0, sizeof(gamma[w]));
	}
	memset(nzHat, 0, sizeof(nzHat));

	// This is where original run method starts
	int j = 0;
	for (j = 0; j < (int) docVector.size(); j++) {
		Document doc = docVector[j];
		for (map<int, int>::iterator iter = doc.termDict.begin(); iter != doc.termDict.end(); iter++) {
			int term = iter->first;
			int k = 0;
			for (k = 0; k < K; k++) {
				gamma[term][k] = ((nPhi[term][k] + eta) / (nz[k] + eta * miniBatch.M)) * (nTheta[doc.docId][k] + alpha);

				nTheta[doc.docId][k] = ((pow((1 - rhoTheta), doc.termDict[term]) * nTheta[doc.docId][k])
						+ ((1 - pow((1 - rhoTheta), doc.termDict[term])) * doc.Cj * gamma[term][k]));
			}
		}

		for (map<int, int>::iterator iter = doc.termDict.begin(); iter != doc.termDict.end(); ++iter) {
			int term = iter->first;
			int k = 0;
			for (k = 0; k < K; k++) {
				gamma[term][k] = ((nPhi[term][k] + eta)	/ (nz[k] + eta * miniBatch.M)) * (nTheta[doc.docId][k] + alpha);

				nTheta[doc.docId][k] = ((pow((1 - rhoTheta), doc.termDict[term]) * nTheta[doc.docId][k])
						+ ((1 - pow((1 - rhoTheta), doc.termDict[term])) * doc.Cj * gamma[term][k]));

				nPhiHat[term][k] += nPhiHat[term][k] + (C * gamma[term][k]/ miniBatch.M);
				nzHat[k] += nzHat[k] + (C * gamma[term][k]/ miniBatch.M);
			}
		}
	}

	for (int k = 0; k < K; k++) {
		for (int w = 1; w < W + 1; w++) {
			nPhi[w][k] = ((1 - rhoPhi) * nPhi[w][k]) + (rhoPhi * nPhiHat[w][k]);
		}
		nz[k] = ((1 - rhoPhi) * nz[k]) + (rhoPhi * nzHat[k]);
	}
	for (int w = 0; w < W + 1; w++) {
		delete[](gamma[w]);
		delete[](nPhiHat[w]);
	}
	delete[](nPhiHat);
	delete[](gamma);
	delete[](nzHat);
}
